### PR TITLE
network: replace detect_container() with udevd_is_active()

### PR DESCRIPTION
**Pull request by yuwata**
Not sure, but may help #7293, https://github.com/coreos/bugs/issues/992, and https://bugs.debian.org/cgi-bin/bugreport.cgi?bug=917128.

I will try to add a testcase for that. Please do not merge this yet. Of course, any comments are welcome.

### CODE REVIEW

**Comment by poettering** [comment]
So, udev *is* conditioned out in containers, the `ConditionPathIsReadWrite=/sys` line in its unit file is about that. The assumption is that container mgrs mount that dir read-only.

See:
https://www.freedesktop.org/wiki/Software/systemd/ContainerInterface/

The second line refers to that (though vaguely)

**Comment by CameronNemo** [comment]
@poettering LXD has supported uevent injection in containers for 5 releases, with the intention of running udev in these containers:

>On very recent kernels, containers that have USB devices setup in LXD will now get add/remove and bind/unbind uevents forwarded to them, allowing for the use of udev rules and other software that listen for uevents.

https://discuss.linuxcontainers.org/t/lxd-3-6-has-been-released/2923

**Comment by yuwata** [comment]
Thank you for the comments.

So, if I understand correctly, we can expect that nothing will be broken by replacing `detect_container()` with `udevd_is_active()` or `udev_queue_get_udev_is_active()` in `networkd`, and hopefully that may make better support on some setups of LXD.

**Comment by yuwata** [comment]
v2: replace `detect_container()` with `udevd_is_active()`.

**Comment by yuwata** [comment]
Hmm, https://github.com/systemd/systemd/blob/c20db3887569e0c0d9c0e2845c5286e7edf0133a/src/libsystemd/sd-device/device-monitor.c#L431-L432

**Comment by poettering** [comment]
> @poettering LXD has supported uevent injection in containers for 5 releases, with the intention of running udev in these containers:

Well, the linux kernel doesn't virtualize any devices besides network devices for containers. This can't work.

For this to work there would have to be a concept of clear ownership concept for devices, i.e. that a device is either available on the host, or on in a container, and not all udevs probe the same devices reconfiguring them in parallel and so on. For that sysfs would have to be changed to only show devices currently owned by the specific environment. Nothing of that exists though. The kernel community refused in the past adding any such infrastructure to the kernel (besides for networking devices), and this hasn't changed so far.

Yes, some container envs ignore that, and pretend sysfs and device ownership wouldn't matter. But this is nothing we can support in udevd. If they break it, they get to keep the pieces. Unless the kernel folks explicitly support device namespacing we can't fake that in userspace, sorry.

**Comment by poettering** [comment]
> So, if I understand correctly, we can expect that nothing will be broken by replacing `detect_container()` with `udevd_is_active()` or `udev_queue_get_udev_is_active()` in `networkd`, and hopefully that may make better support on some setups of LXD.

This sounds racy, no? networkd only has an ordering dep against udev because of the tuntap stuff (see 4bd5ace3e78ec0ca4c174bc875e3d9f6e1ae7405), and the intention was to remove that. With that removed this would become racy no? (i mean, we can decide that the After= thing has to stay, but if so we should do that for good reasons)

File: src/network/networkd-manager.c [hunk 1]
```diff
@@ -205,6 +205,15 @@ static int manager_udev_process_link(sd_device_monitor *monitor, sd_device *devi
...
         }
 
+        r = sd_device_get_is_initialized(device);
+        if (r < 0) {
+                log_device_debug_errno(device, r, "Could not determine whether the device is initialized or not, ignoring uevent: %m");
+                return 0;
+        } else if (r == 0) {
+                log_device_debug(device, "Device is not initialized yet, ignoring uevent.");
+                return 0;
+        }
```
Comment (poettering) [inline src/network/networkd-manager.c:215]: i am not sure i grok this change. why would we check this here? i mean, whether a device is initialized or not basically just means that the uevent for the device popping up has been sent out, hence this check here should be redundant, no?

**Comment by yuwata** [comment]
@poettering Thank you for the comments and also in that in telegram.

> > So, if I understand correctly, we can expect that nothing will be broken by replacing `detect_container()` with `udevd_is_active()` or `udev_queue_get_udev_is_active()` in `networkd`, and hopefully that may make better support on some setups of LXD.
> 
> This sounds racy, no? networkd only has an ordering dep against udev because of the tuntap stuff (see [4bd5ace](https://github.com/systemd/systemd/commit/4bd5ace3e78ec0ca4c174bc875e3d9f6e1ae7405)), and the intention was to remove that. With that removed this would become racy no? (i mean, we can decide that the After= thing has to stay, but if so we should do that for good reasons)

Indeed...

At least, this improves or fixes nothing, though, I am not sure about the LXD case. Let's close this.

**Pull request closed without merging**